hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/Router.java

```diff
@@ -386,6 +380,26 @@ public void run() {
   // RPC Server
   /////////////////////////////////////////////////////////
 
+  private void setRpcServer() throws IOException {
+    if (conf.getBoolean(
+            RBFConfigKeys.DFS_ROUTER_RPC_ENABLE,
+            RBFConfigKeys.DFS_ROUTER_RPC_ENABLE_DEFAULT)) {
+      // Create RPC server
+      this.rpcServer = createRpcServer();
+      addService(this.rpcServer);
+      this.setRpcServerAddress(rpcServer.getRpcAddress());
+    } else {
+      InetSocketAddress confRpcAddress = conf.getSocketAddr(
+              RBFConfigKeys.DFS_ROUTER_RPC_BIND_HOST_KEY,
+              RBFConfigKeys.DFS_ROUTER_RPC_ADDRESS_KEY,
+              RBFConfigKeys.DFS_ROUTER_RPC_ADDRESS_DEFAULT,
+              RBFConfigKeys.DFS_ROUTER_RPC_PORT_DEFAULT);
+
+      String hostname = InetAddress.getLocalHost().getHostName();
```

Comment:
Most of this is already defined in setRpcServerAddress() and it even considers exception, etc.
I would refactor that function to allow specifying a customized port.

In any case, I would leave the general code as it was, and then after that, check if routerId is null,  and then set it to this.

Reply:
Thanks for your advice. The method setRpcServerAddress() is only called when rpc.enable is true. Maybe we can add a new method after that and check routerid.